src/GraphWebhooks/Controllers/ListenController.cs

```diff
@@ -0,0 +1,215 @@
+                    // Add notification to list to process. If there is more than
+                    // one notification for a given resource, we'll only process it once
+                    messageNotifications[notification.Resource] = notification;
+                }
+            }
+
+            // Since resource data is not included in these notifications,
+            // use Microsoft Graph to get the messages
+            await GetMessagesAsync(messageNotifications.Values);
+
+            // Process encrypted notifications
+            var clientNotifications = new List<ClientNotification>();
+            foreach (var notification in notifications.Value.Where(n => n.EncryptedContent != null))
+            {
+                // Decrypt the encrypted payload using private key
+                var chatMessage = await notification.EncryptedContent.DecryptAsync<ChatMessage>(async (id, thumbprint) => {
+                    return await _certificateService.GetDecryptionCertificate();
+                });
+
+                // Add a SignalR notification for this message to the list
+                clientNotifications.Add(new ClientNotification(new
+                {
+                    Sender = chatMessage.From?.User?.DisplayName ?? "UNKNOWN",
+                    Message = chatMessage.Body?.Content ?? ""
+                }));
+            }
+
+            // Send SignalR notifications
+            if (clientNotifications.Count > 0)
+            {
+                await _hubContext.Clients.All.SendAsync("showNotification", clientNotifications);
+            }
+
+            // Return 202 to Graph to confirm receipt of notification.
+            // Not sending this will cause Graph to retry the notification.
+            return Accepted();
+        }
+
+        /// <summary>
+        /// Validates tokens contained in a ChangeNotificationCollection
+        /// </summary>
+        /// <param name="notifications">The ChangeNotificationCollection to validate</param>
+        /// <param name="tenantIds">A list of expected tenant IDs</param>
+        /// <param name="appIds">A list of expected app IDs</param>
+        /// <returns>true if all tokens are valid, false if not</returns>
+        private async Task<bool> AreTokensValid(
+            ChangeNotificationCollection notifications,
+            List<Guid> tenantIds,
+            List<Guid> appIds)
+        {
+            // First validate assuming tokens are v2 tokens
+            bool areTokensValid = await notifications.AreTokensValidV2(tenantIds, appIds);
+
+            if (!areTokensValid)
+            {
+                // If v2 validation failed, try v1
+                // This method throws if validation fails, so catch any exception
+                // and treat as validation failure
+                try
+                {
+                    areTokensValid = await notifications.AreTokensValid(tenantIds, appIds);
+                }
+                catch
```

Comment:
maybe specialize the exception type you're expecting, so we don't mask other unexpected exceptions